bin/provision-instances

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/python3
+#
+# This script runs a full instance provision over all non-ephemeral instances
+# on OCIM. Running this script requires Django to be initialized - the easiest
+# way to do this is by evaluating the script from within the Django shell by
+# running:
+#
+#   > eval(open('bin/provision-instances').open())
+#
+# The provision process happens in "waves", where all instances that fail to
+# provision in a given wave are recorded and reprovisioned during the next wave.
+#
+# This provision script is designed to withstand restarts of either OCIM or
+# this script itself; it accomplishes this by always saving and loading the
+# state of the provision process to/from `ocim_update_waves.ini`.
+
+import ast
+from configparser import ConfigParser
+from datetime import datetime, timedelta
+import os
+import pytz
+from time import sleep
+
+from instance.models.openedx_instance import OpenEdXInstance
+from instance.tasks import spawn_appserver
+
+WAVES_FILE = 'ocim_update_waves.ini'
+SLEEP_TIME = 60
+MAX_PROCESSING = 6
+MAX_PROCESSING_TIME = 240  # 4 hours
+
+waves_config = ConfigParser()
+
+
+def bootstrap_instance_list():
+    # Collect all of the instances that should be included in the OCIM upgrades.
+    #
+    # Instances with any of the following attributes at the start of the upgrade
+    # are avoided for the entire upgrade:
+    #   - Any instance considered "shut down" (all VMs terminated)
+    #   - Any instance without any appservers
+    #   - Any instance set to use ephemeral databases
+    #   - Any instance where the latest appserver isn't in a healthy stable state (e.g. running)
+
+    failed_results = []
+    ready_instances = []
+    for inst in OpenEdXInstance.objects.all():
+        # Set field defaults for ALL instances, even ones we're not going to reprovision.
+        inst.set_field_defaults()
+        inst.save()
+
+        latest_appserver = inst.appserver_set.last()
+
+        if inst.is_shut_down:
+            print('-', end='', flush=True)
+            continue
+
+        if not latest_appserver:
+            failed_results.append((inst, 'is has no appservers'))
+            print('n', end='', flush=True)
+            continue
+
+        if inst.use_ephemeral_databases:
+            failed_results.append((inst, 'it has "use_ephemeral_databases" enabled'))
+            print('e', end='', flush=True)
+            continue
+
+        if not (latest_appserver.status.is_steady_state and latest_appserver.status.is_healthy_state):
+            failed_results.append((inst, 'it\'s latest appserver is in the "{}" state'.format(
+                latest_appserver.status.name
+            )))
```

Comment:
If the latest appserver isn't healthy, that means it doesn't have a configuration which will succeed when we try to redeploy it. Later on if/when someone fixes its configuration and redeploys/activates an appserver, it'll be redeployed at that point.